File: annotation-processor/src/main/java/me/bristermitten/mittenlib/annotations/config/ConfigProcessor.java
package me.bristermitten.mittenlib.annotations.config;

import com.google.auto.service.AutoService;
import com.google.inject.Guice;
import com.palantir.javapoet.JavaFile;
import io.toolisticon.aptk.common.ToolingProvider;
import io.toolisticon.aptk.tools.AbstractAnnotationProcessor;
import io.toolisticon.aptk.tools.MessagerUtils;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Set;
import javax.annotation.processing.Processor;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedAnnotationTypes;
import javax.annotation.processing.SupportedSourceVersion;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.NestingKind;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.ElementFilter;
import me.bristermitten.mittenlib.annotations.ast.AbstractConfigStructure;
import me.bristermitten.mittenlib.annotations.compile.ConfigImplGenerator;
import me.bristermitten.mittenlib.annotations.compile.ConfigLoaderGenerator;
import me.bristermitten.mittenlib.annotations.compile.ConfigLoaderModuleGenerator;
import me.bristermitten.mittenlib.annotations.compile.ConfigProcessorModule;
import me.bristermitten.mittenlib.annotations.compile.ConfigSaverGenerator;
import me.bristermitten.mittenlib.annotations.compile.ConfigValidatorGenerator;
import me.bristermitten.mittenlib.annotations.compile.ConfigurationClassNameGenerator;
import me.bristermitten.mittenlib.annotations.compile.NewtypeImplGenerator;
import me.bristermitten.mittenlib.annotations.exception.ConfigProcessingException;
import me.bristermitten.mittenlib.annotations.parser.ASTVerifier;
import me.bristermitten.mittenlib.annotations.parser.ConfigClassParser;
import me.bristermitten.mittenlib.annotations.parser.CustomDeserializers;
import me.bristermitten.mittenlib.annotations.parser.CustomSerializers;
import me.bristermitten.mittenlib.config.Config;
import me.bristermitten.mittenlib.config.Newtype;
import me.bristermitten.mittenlib.config.extension.CustomDeserializerFor;
import me.bristermitten.mittenlib.config.extension.CustomSerializerFor;

/**
 * Annotation processor for generating configuration classes from DTO classes marked with {@link
 * Config}. This processor handles the compilation-time generation of implementation classes for
 * configuration DTOs, creating strongly typed configuration objects with proper getters, equals,
 * hashCode, and toString methods. The processor only processes top-level classes (not nested
 * classes).
 */
@SupportedAnnotationTypes({"me.bristermitten.mittenlib.config.Config", "me.bristermitten.mittenlib.config.Newtype"})
@SupportedSourceVersion(SourceVersion.RELEASE_21)
@AutoService(Processor.class)
public class ConfigProcessor extends AbstractAnnotationProcessor {

    /** Public constructor for the compiler */
    public ConfigProcessor() {
        super();
    }

    /**
     * Processes annotations and generates configuration implementation classes. This method is called
     * by the Java compiler during the annotation processing phase. It performs the following steps:
     * 1. Sets up the tooling environment and creates a Guice injector 2. Finds all top-level classes
     * annotated with @Config 3. Parses each class into an abstract configuration structure 4.
     * Generates implementation classes for each structure 5. Writes the generated files to the filer
     *
     * @param annotations The annotation types requested to be processed
     * @param roundEnv The environment for this round of annotation processing
     * @return true if the annotations were processed successfully, false otherwise
     * @throws ConfigProcessingException if there is an error writing the generated files
     */
    @Override
    public boolean processAnnotations(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
        ToolingProvider.setTooling(processingEnv);
        var injector = Guice.createInjector(new ConfigProcessorModule(processingEnv));

        final List<TypeElement> types = roundEnv.getElementsAnnotatedWith(Config.class).stream()
                .filter(TypeElement.class::isInstance)
                .map(TypeElement.class::cast)
                .filter(element -> element.getNestingKind() == NestingKind.TOP_LEVEL)
                .toList();

        final List<TypeElement> newtypes = roundEnv.getElementsAnnotatedWith(Newtype.class).stream()
                .filter(TypeElement.class::isInstance)
                .map(TypeElement.class::cast)
                .toList();

        boolean anyErrors = false;

        var newtypeGenerator = new NewtypeImplGenerator();
        for (TypeElement newtype : newtypes) {
            if (newtype.getKind() != ElementKind.RECORD && newtype.getKind() != ElementKind.INTERFACE) {
                MessagerUtils.error(
                        newtype,
                        "Newtype annotation only supports records and interfaces: " + newtype.getQualifiedName());
                anyErrors = true;
                continue;
            }

            if (newtype.getKind() == ElementKind.RECORD) {
                if (newtype.getRecordComponents().size() != 1) {
                    MessagerUtils.error(
                            newtype,
                            "Newtype record " + newtype.getQualifiedName() + " must have exactly one component");
                    anyErrors = true;
                }
            } else if (newtype.getKind() == ElementKind.INTERFACE) {
                List<ExecutableElement> methods = ElementFilter.methodsIn(newtype.getEnclosedElements()).stream()
                        .filter(m -> !m.isDefault() && !m.getModifiers().contains(Modifier.STATIC))
                        .toList();
                if (methods.size() != 1) {
                    MessagerUtils.error(
                            newtype,
                            "Newtype interface " + newtype.getQualifiedName()
                                    + " must have exactly one abstract method");
                    anyErrors = true;
                    continue;
                }

                if (!anyErrors) {
                    JavaFile emit = newtypeGenerator.emit(newtype);
                    try {
                        emit.writeTo(processingEnv.getFiler());
                    } catch (Exception e) {
                        throw new ConfigProcessingException("Could not create newtype impl file", e);
                    }
                }
            }
        }

        CustomDeserializers customDeserializers = injector.getInstance(CustomDeserializers.class);
        roundEnv.getElementsAnnotatedWith(CustomDeserializerFor.class).stream()
                .map(TypeElement.class::cast)
                .forEach(customDeserializers::registerCustomDeserializer);

        CustomSerializers customSerializers = injector.getInstance(CustomSerializers.class);
        roundEnv.getElementsAnnotatedWith(CustomSerializerFor.class).stream()
                .map(TypeElement.class::cast)
                .forEach(customSerializers::registerCustomSerializer);

        List<AbstractConfigStructure> asts = new ArrayList<>();
        var configClassParser = injector.getInstance(ConfigClassParser.class);
        for (TypeElement clazz : types) {
            var ast = configClassParser.parseAbstract(clazz);
            asts.add(ast);
        }

        ASTVerifier verifier = injector.getInstance(ASTVerifier.class);
        for (AbstractConfigStructure ast : asts) {
            if (!verifier.verify(ast)) {
                anyErrors = true;
            }
        }

        if (anyErrors) {
            return true;
        }

        var generator = injector.getInstance(ConfigImplGenerator.class);
        var loaderGenerator = injector.getInstance(ConfigLoaderGenerator.class);
        var saverGenerator = injector.getInstance(ConfigSaverGenerator.class);
        var validatorGenerator = injector.getInstance(ConfigValidatorGenerator.class);
        for (AbstractConfigStructure ast : asts) {
            JavaFile emit = generator.emit(ast);
            JavaFile loaderEmit = loaderGenerator.emit(ast);
            JavaFile saverEmit = saverGenerator.emit(ast);
            JavaFile validatorEmit = validatorGenerator.emit(ast);
            try {
                emit.writeTo(processingEnv.getFiler());
                loaderEmit.writeTo(processingEnv.getFiler());
                saverEmit.writeTo(processingEnv.getFiler());
                validatorEmit.writeTo(processingEnv.getFiler());
            } catch (Exception e) {
                throw new ConfigProcessingException("Could not create config file", e);
            }
        }

        if (!asts.isEmpty()) {
            var moduleGenerator = injector.getInstance(ConfigLoaderModuleGenerator.class);
            var classNameGenerator = injector.getInstance(ConfigurationClassNameGenerator.class);

            // Sort by package name and then simple name for stability
            asts.sort(Comparator.comparing((AbstractConfigStructure ast) ->
                            classNameGenerator.getPublicClassName(ast).packageName())
                    .thenComparing(
                            ast -> classNameGenerator.getPublicClassName(ast).simpleName()));

            // Use the shortest package name as the "root" package for the module
            String rootPackage = asts.stream()
                    .map(ast -> classNameGenerator.getPublicClassName(ast).packageName())
                    .min(Comparator.comparingInt(String::length))
                    .orElse("");

            JavaFile moduleEmit = moduleGenerator.emit(asts, rootPackage);
            try {
                moduleEmit.writeTo(processingEnv.getFiler());
            } catch (Exception e) {
                throw new ConfigProcessingException("Could not create ConfigLoaderModule file", e);
            }
        }
        return true;
    }
}

File: annotation-processor/src/main/scala/me/bristermitten/mittenlib/annotations/config/ConfigProcessor.scala
package me.bristermitten.mittenlib.annotations.config

import com.google.auto.service.AutoService
import com.google.inject.Guice
import com.palantir.javapoet.JavaFile
import io.toolisticon.aptk.common.ToolingProvider
import io.toolisticon.aptk.tools.AbstractAnnotationProcessor
import io.toolisticon.aptk.tools.MessagerUtils
import javax.annotation.processing.Processor
import javax.annotation.processing.RoundEnvironment
import javax.annotation.processing.SupportedAnnotationTypes
import javax.annotation.processing.SupportedSourceVersion
import javax.lang.model.SourceVersion
import javax.lang.model.element.ElementKind
import javax.lang.model.element.NestingKind
import javax.lang.model.element.TypeElement
import me.bristermitten.mittenlib.annotations.compile.ConfigImplGenerator
import me.bristermitten.mittenlib.annotations.compile.ConfigLoaderGenerator
import me.bristermitten.mittenlib.annotations.compile.ConfigLoaderModuleGenerator
import me.bristermitten.mittenlib.annotations.compile.ConfigProcessorModule
import me.bristermitten.mittenlib.annotations.compile.ConfigSaverGenerator
import me.bristermitten.mittenlib.annotations.compile.ConfigValidatorGenerator
import me.bristermitten.mittenlib.annotations.compile.ConfigurationClassNameGenerator
import me.bristermitten.mittenlib.annotations.compile.NewtypeImplGenerator
import me.bristermitten.mittenlib.annotations.domain.ConfigStructure
import me.bristermitten.mittenlib.annotations.exception.ConfigProcessingException
import me.bristermitten.mittenlib.annotations.util.NewtypeUtil
import me.bristermitten.mittenlib.annotations.parser.ConfigParser
import me.bristermitten.mittenlib.annotations.parser.CustomDeserializers
import me.bristermitten.mittenlib.annotations.parser.CustomSerializers
import me.bristermitten.mittenlib.annotations.parser.ParserError
import me.bristermitten.mittenlib.config.Config
import me.bristermitten.mittenlib.config.Newtype
import me.bristermitten.mittenlib.config.extension.CustomDeserializerFor
import me.bristermitten.mittenlib.config.extension.CustomSerializerFor
import scala.jdk.CollectionConverters.*

import cats.data.ValidatedNel
import cats.implicits.*

/** Annotation processor for generating configuration classes from DTO classes
  * marked with {@link Config}.
  */
@SupportedAnnotationTypes(
  Array(
    "me.bristermitten.mittenlib.config.Config",
    "me.bristermitten.mittenlib.config.Newtype"
  )
)
@SupportedSourceVersion(SourceVersion.RELEASE_21)
@AutoService(Array(classOf[Processor]))
class ConfigProcessor extends AbstractAnnotationProcessor:

  override def processAnnotations(
      annotations: java.util.Set[? <: TypeElement],
      roundEnv: RoundEnvironment
  ): Boolean =
    ToolingProvider.setTooling(processingEnv)
    val injector = Guice.createInjector(ConfigProcessorModule(processingEnv))

    // 1. Setup side-effecting registrations
    val customDeserializers = injector.getInstance(classOf[CustomDeserializers])
    roundEnv
      .getElementsAnnotatedWith(classOf[CustomDeserializerFor])
      .asScala
      .collect { case t: TypeElement => t }
      .foreach(customDeserializers.registerCustomDeserializer)

    val customSerializers = injector.getInstance(classOf[CustomSerializers])
    roundEnv
      .getElementsAnnotatedWith(classOf[CustomSerializerFor])
      .asScala
      .collect { case t: TypeElement => t }
      .foreach(customSerializers.registerCustomSerializer)

    val types = roundEnv
      .getElementsAnnotatedWith(classOf[Config])
      .asScala
      .collect { case t: TypeElement => t }
      .filter(_.getNestingKind == NestingKind.TOP_LEVEL)
      .toList

    val newtypes = roundEnv
      .getElementsAnnotatedWith(classOf[Newtype])
      .asScala
      .collect { case t: TypeElement => t }
      .toList

    val configParser = injector.getInstance(classOf[ConfigParser])

    // 2. Functional compilation/parsing pipeline using Cats ValidatedNel
    val newtypeResults: ValidatedNel[ParserError, List[TypeElement]] =
      newtypes.traverse(validateNewtype)

    val configResults: ValidatedNel[ParserError, List[ConfigStructure]] =
      types.traverse(configParser.parseAbstract)

    (newtypeResults, configResults).mapN((_, _)) match
      case cats.data.Validated.Invalid(errors) =>
        errors.toList.foreach(error =>
          MessagerUtils.error(error.element, error.message)
        )
        true

      case cats.data.Validated.Valid((validNewtypes, validConfigs)) =>
        // 3. Code generation
        val newtypeGenerator = NewtypeImplGenerator()
        val generator = injector.getInstance(classOf[ConfigImplGenerator])
        val loaderGenerator =
          injector.getInstance(classOf[ConfigLoaderGenerator])
        val saverGenerator = injector.getInstance(classOf[ConfigSaverGenerator])
        val validatorGenerator =
          injector.getInstance(classOf[ConfigValidatorGenerator])
        val moduleGenerator =
          injector.getInstance(classOf[ConfigLoaderModuleGenerator])
        val classNameGenerator =
          injector.getInstance(classOf[ConfigurationClassNameGenerator])

        val filesToWrite = List.newBuilder[JavaFile]

        // Emit interfaces implementation newtype classes
        validNewtypes
          .filter(_.getKind == ElementKind.INTERFACE)
          .foreach(nt => filesToWrite += newtypeGenerator.emit(nt))

        // Emit config implementation, loaders, savers, and validators
        validConfigs.foreach { ast =>
          filesToWrite += generator.emit(ast)
          filesToWrite += loaderGenerator.emit(ast)
          filesToWrite += saverGenerator.emit(ast)
          filesToWrite += validatorGenerator.emit(ast)
        }

        // Emit loader module
        if (validConfigs.nonEmpty) {
          val sortedConfigs = validConfigs.sortBy { ast =>
            val pubClass = classNameGenerator.getPublicClassName(ast)
            (pubClass.packageName(), pubClass.simpleName())
          }

          val rootPackage = sortedConfigs
            .map(ast =>
              classNameGenerator.getPublicClassName(ast).packageName()
            )
            .minBy(_.length)

          filesToWrite += moduleGenerator.emit(
            sortedConfigs.asJava,
            rootPackage
          )
        }

        // 4. Write files to Filer
        try filesToWrite.result().foreach(_.writeTo(processingEnv.getFiler))
        catch
          case e: Exception =>
            throw ConfigProcessingException(
              "Could not write generated config files",
              e
            )
        true

  private def validateNewtype(
      newtype: TypeElement
  ): ValidatedNel[ParserError, TypeElement] =
    val errors = List.newBuilder[ParserError]
    if (
      newtype.getKind != ElementKind.RECORD && newtype.getKind != ElementKind.INTERFACE
    ) {
      errors += ParserError(
        newtype,
        s"Newtype annotation only supports records and interfaces: ${newtype.getQualifiedName}"
      )
    } else if (newtype.getKind == ElementKind.RECORD) {
      if (newtype.getRecordComponents.size() != 1) {
        errors += ParserError(
          newtype,
          s"Newtype record ${newtype.getQualifiedName} must have exactly one component"
        )
      }
    } else if (newtype.getKind == ElementKind.INTERFACE) {
      val methods = NewtypeUtil.getAbstractMethods(newtype)
      if (methods.size != 1) {
        errors += ParserError(
          newtype,
          s"Newtype interface ${newtype.getQualifiedName} must have exactly one abstract method"
        )
      }
    }

    val errList = errors.result()
    if (errList.isEmpty) {
      newtype.validNel
    } else {
      cats.data.Validated.invalid(
        cats.data.NonEmptyList.fromListUnsafe(errList)
      )
    }
